runtime/buffer.h

```diff
@@ -422,64 +423,75 @@ void for_each_tile(const Tile& tile, raw_buffer& buf, F&& f) {
 // This function is not fast, use it for non-performance critical code. It is useful for
 // making rank-agnostic algorithms without a recursive wrapper, which is otherwise difficult.
 template <typename F>
-void for_each_index(span<const dim> dims, F&& f) {
+void for_each_index(span<const dim> dims, const F& f) {
   // Not using alloca for performance, but to avoid including <vector>
   index_t* i = SLINKY_ALLOCA(index_t, dims.size());
   internal::for_each_index(dims, dims.size() - 1, i, dims.size(), f);
 }
 template <typename F>
-void for_each_index(const raw_buffer& buf, F&& f) {
+void for_each_index(const raw_buffer& buf, const F& f) {
   for_each_index({buf.dims, buf.rank}, f);
 }
 
 // Call `f(void* base, index_t extent)` for each contiguous slice in the domain of `buf`.
 // This function attempts to be efficient to support production quality implementations of callbacks.
 template <typename F>
-void for_each_contiguous_slice(const raw_buffer& buf, F&& f) {
+void for_each_contiguous_slice(const raw_buffer& buf, const F& f) {
   internal::for_each_contiguous_slice(buf.base, buf.dims, buf.rank - 1, buf.elem_size, f);
 }
 
 // Call `f` for each slice of the first `slice_rank` dimensions of buf.
 template <typename F>
-void for_each_slice(std::size_t slice_rank, const raw_buffer& const_buf, F&& f) {
+void for_each_slice(std::size_t slice_rank, const raw_buffer& const_buf, const F& f) {
   // TODO: Should we make a copy of the buffer here? We const_cast it so we can modify it, but we
   // also restore it to its original state... not thread safe though in case buf is being shared
   // with another thread.
   raw_buffer& buf = const_cast<raw_buffer&>(const_buf);
```

Comment:
const-casting this away seems like a really scary idea -- the caller expects it not to get mutated. IMHO we should definitely make a copy.

Reply:
I did it for now, but I think we might go back to const_cast at some point. We'll see if the overhead/stack usage matters. I agree it's better to copy if we can afford it.